src/profile-logic/profile-data.js

```diff
@@ -1001,6 +1003,44 @@ export function getTracingMarkers(thread: Thread): TracingMarker[] {
     } else if ('startTime' in data && 'endTime' in data) {
       const { startTime, endTime } = data;
       if (typeof startTime === 'number' && typeof endTime === 'number') {
+        if (data.type === 'UserTiming') {
+          if (data.entryType === 'measure') {
+            // De-duplicate performance.mark, as performance.measure can be built from
+            // marks.
+            const { startMark, endMark } = data;
+
+            // Check start marks.
+            if (typeof startMark === 'string') {
+              const startMarkIndex = markNameToTracingIndex.get(startMark);
+              if (startMarkIndex === undefined) {
+                // This mark has not been found yet, remember it.
+                markNamesToSkip.add(startMark);
+              } else {
+                // This mark was already added as a tracing marker, remove it now.
+                tracingMarkers[startMarkIndex] = null;
+              }
+            }
+
+            // Check end marks.
+            if (typeof endMark === 'string') {
+              const endMarkIndex = markNameToTracingIndex.get(endMark);
+              if (endMarkIndex === undefined) {
+                // This mark has not been found yet, remember it.
+                markNamesToSkip.add(endMark);
+              } else {
+                // This mark was already added as a tracing marker, remove it now.
+                tracingMarkers[endMarkIndex] = null;
+              }
+            }
+          } else {
+            const { name } = data;
+            markNameToTracingIndex.set(name, tracingMarkers.length);
+            if (markNamesToSkip.has(name)) {
+              markNamesToSkip.delete(name);
+              continue;
+            }
+          }
+        }
```

Comment:
The algorithm works fine for simple cases, but has bugs for at least 2 cases involving markers holding the same names. 

You can [use this profile directly online](http://localhost:4242/public/70c4b35ae783d73482356213d71d46c320549e56/marker-chart/?hiddenThreads=0-1-2-3&range=1.7857_2.6630&react_perf&thread=4&threadOrder=0-2-3-4-1&v=3) or capture from https://bug1425605.bmoattachments.org/attachment.cgi?id=8958471 (you need to start capturing and _then_ press f5 as the test runs all cases at load time).

See comments below, explaining the bugs
```javascript
// timeout is a function that returns a promise resolving after a timeout.
async function testcase7() {
  performance.mark('duplicate7-start');
  await timeout(50);
  performance.mark('duplicate7-end');
  await timeout(50);
  performance.mark('duplicate7-end');
  await timeout(50);
  performance.measure('duplicate7', 'duplicate7-start', 'duplicate7-end');
  // KO the wrong end marker is deleted.
}
```
Indeed `measure` uses the most recent marker with the specified name.
But when capturing, the marker order we get is:
* [measure] duplicate7
* [mark] duplicate7-start
* [mark] duplicate7-end
* [mark] duplicate7-end

So the first marker with the name `duplicate7-end` is found first in the algorithm and is deleted, but the last one should be deleted because this is the one used in the `measure` marker.

```javascript
async function testcase8() {
  performance.mark('duplicate8-start');
  await timeout(50);
  performance.measure('duplicate8', 'duplicate8-start');
  await timeout(50);
  performance.mark('duplicate8-start');
  await timeout(50);
  performance.measure('duplicate8', 'duplicate8-start');
  // KO the start marker isn't deleted.
}
```

This is maybe related to the aforementioned issue in gecko.

The problem here is that the "measure" marker always appears before the start "mark" marker.
So the order we get here is:
* [measure] duplicate8
* [mark] duplicate8-start
* [measure] duplicate8
* [mark] duplicate8-start

When handling the second measure `duplicate8` (note: the measure name isn't important here), in `markNameToTracingIndex` we still have the reference to the first `duplicate8-start` for that name, and as a result that one is being nullified (note: it already was).
Only then we encounter the second marker but because it's not been added to `markNamesToSkip` it's displayed.

```javascript
async function testcase9() {
  performance.mark('duplicate9-start');
  await timeout(50);
  performance.mark('duplicate9-start');
  await timeout(50);
  performance.measure('duplicate9', 'duplicate9-start');
  // KO The wrong marker is deleted.
```

The reason is the same one as in testcase8, but the outcome is more important: we deleted a marker that shouldn't have been deleted.